kv/kvclient/kvcoord: TestDistSenderReplicaStall failed

kv/kvclient/kvcoord.TestDistSenderReplicaStall [failed](https://tanzanite.cluster.engflow.com/invocations/default/44226caa-e9a9-4620-90b8-b9e5ce0f4ac1?testReportRun=9&testReportShard=7&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZjbGllbnQva3Zjb29yZDprdmNvb3JkX3Rlc3Q=) on release-25.1.6-rc @ [8277f164925ba5fb24d987ff1af4703d5f084da1](https://github.com/cockroachdb/cockroach/commits/8277f164925ba5fb24d987ff1af4703d5f084da1):


```
        	            	  | github.com/cockroachdb/cockroach/pkg/testutils.RunValues[...].func1
        	            	  | 	pkg/testutils/subtest.go:23
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1689
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1695
        	            	Wraps: (2) aborted in DistSender
        	            	Wraps: (3) result is ambiguous: context deadline exceeded
        	            	Wraps: (4) secondary error attachment
        	            	  | aborted during DistSender.Send: context deadline exceeded
        	            	  | (1)
        	            	  |   | (opaque error wrapper)
        	            	  |   | type name: github.com/cockroachdb/errors/withstack/*withstack.withStack
        	            	  |   | reportable 0:
        	            	  |   |
        	            	  |   | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).sendToReplicas
        	            	  |   | 	pkg/kv/kvclient/kvcoord/dist_sender.go:3124
        	            	  |   | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).sendPartialBatch
        	            	  |   | 	pkg/kv/kvclient/kvcoord/dist_sender.go:2211
        	            	  |   | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).divideAndSendBatchToRanges
        	            	  |   | 	pkg/kv/kvclient/kvcoord/dist_sender.go:1746
        	            	  |   | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).Send
        	            	  |   | 	pkg/kv/kvclient/kvcoord/dist_sender.go:1265
        	            	  |   | github.com/cockroachdb/cockroach/pkg/kv.(*CrossRangeTxnWrapperSender).Send
        	            	  |   | 	pkg/kv/db.go:224
        	            	  |   | github.com/cockroachdb/cockroach/pkg/kv.(*DB).sendUsingSender
        	            	  |   | 	pkg/kv/db.go:1159
        	            	  |   | github.com/cockroachdb/cockroach/pkg/kv.(*DB).send
        	            	  |   | 	pkg/kv/db.go:1142
        	            	  |   | github.com/cockroachdb/cockroach/pkg/kv.sendAndFill
        	            	  |   | 	pkg/kv/db.go:951
        	            	  |   | github.com/cockroachdb/cockroach/pkg/kv.(*DB).Run
        	            	  |   | 	pkg/kv/db.go:974
        	            	  |   | github.com/cockroachdb/cockroach/pkg/kv.(*DB).Inc
        	            	  |   | 	pkg/kv/db.go:485
        	            	  |   | github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord_test.TestDistSenderReplicaStall.func1
        	            	  |   | 	pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go:82
        	            	  |   | github.com/cockroachdb/cockroach/pkg/testutils.RunValues[...].func1
        	            	  |   | 	pkg/testutils/subtest.go:23
        	            	  |   | testing.tRunner
        	            	  |   | 	GOROOT/src/testing/testing.go:1689
        	            	  |   | runtime.goexit
        	            	  |   | 	src/runtime/asm_amd64.s:1695
        	            	  | Wraps: (2) aborted during DistSender.Send
        	            	  | Wraps: (3) context deadline exceeded
        	            	  | Error types: (1) *errbase.opaqueWrapper (2) *errutil.withPrefix (3) context.deadlineExceededError
        	            	Wraps: (5) context deadline exceeded
        	            	Error types: (1) *errbase.opaqueWrapper (2) *errutil.withPrefix (3) *kvpb.AmbiguousResultError (4) *secondary.withSecondaryError (5) context.deadlineExceededError
        	Test:       	TestDistSenderReplicaStall/clientTimeout=false
    --- FAIL: TestDistSenderReplicaStall/clientTimeout=false (31.31s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=9</code>
 - <code>shard=7</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDistSenderReplicaStall.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-50361

--- a/pkg/kv/kvclient/kvcoord/BUILD.bazel
+++ b/pkg/kv/kvclient/kvcoord/BUILD.bazel
@@ -218,6 +218,7 @@ go_test(
         "//pkg/testutils/testcluster",
         "//pkg/util",
         "//pkg/util/admission/admissionpb",
+        "//pkg/util/allstacks",
         "//pkg/util/caller",
         "//pkg/util/ctxgroup",
         "//pkg/util/encoding",

--- a/pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go
+++ b/pkg/kv/kvclient/kvcoord/dist_sender_circuit_breaker_test.go
@@ -24,6 +24,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
+	"github.com/cockroachdb/cockroach/pkg/util/allstacks"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/stop"
@@ -50,6 +51,9 @@ func TestDistSenderReplicaStall(t *testing.T) {
 
 	testutils.RunTrueAndFalse(t, "clientTimeout", func(t *testing.T, clientTimeout bool) {
 		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+		defer time.AfterFunc(29*time.Second, func() {
+			log.Errorf(ctx, "about to time out, all stacks:\n\n%s", allstacks.Get())
+		}).Stop()
 		defer cancel()
 
 		// The lease won't move unless we use expiration-based leases. We also

